File: sdks/rust/src/teardown.rs
//! Tear down scoped test data via raw SQL DELETE in reverse topological order.

use serde_json::Value;
use std::collections::{HashMap, HashSet};

use crate::dialect::Dialect;
use crate::graph::{find_deferrable_edge, topo_sort};
use crate::types::{SchemaInfo, SqlExecutor};

pub async fn teardown(
    executor: &dyn SqlExecutor,
    dialect: &dyn Dialect,
    table_map: &HashMap<String, String>,
    column_maps: &HashMap<String, HashMap<String, String>>,
    schema: &SchemaInfo,
    scope_value: &str,
    refs: Option<&Value>,
) -> Result<(), String> {
    // Convert edges to Value format for graph module
    let edge_dicts: Vec<Value> = schema
        .edges
        .iter()
        .map(|e| {
            serde_json::json!({
                "from": e.from_model,
                "to": e.to_model,
                "localField": e.local_field,
                "foreignField": e.foreign_field,
                "nullable": e.nullable
            })
        })
        .collect();

    // Find scope root model
    let scope_root_model: Option<String> = schema.edges.iter().find_map(|edge| {
        if edge.local_field.to_lowercase() == schema.scope_field.to_lowercase()
            && edge.to_model != edge.from_model
        {
            Some(edge.to_model.clone())
        } else {
            None
        }
    });

    // Build map: model -> FK field pointing to scope root
    let mut scope_field_by_model: HashMap<String, String> = HashMap::new();
    if let Some(ref root) = scope_root_model {
        for edge in &schema.edges {
            if edge.to_model == *root && edge.from_model != *root {
                scope_field_by_model.insert(edge.from_model.clone(), edge.local_field.clone());
            }
        }
    }

    let model_names: Vec<String> = schema.models.iter().map(|m| m.name.clone()).collect();
    let sort_result = topo_sort(&model_names, &edge_dicts);
    let sorted_models: Vec<String> = sort_result
        .get("sorted")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();
    let cycles: Vec<Vec<String>> = sort_result
        .get("cycles")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| {
                    v.as_array().map(|cycle| {
                        cycle
                            .iter()
                            .filter_map(|n| n.as_str().map(String::from))
                            .collect()
                    })
                })
                .collect()
        })
        .unwrap_or_default();

    // Transaction wrapper — use the executor's transaction method
    // Break cycles by nullifying deferrable FKs
    for cycle in &cycles {
        let cycle_strs: Vec<String> = cycle.clone();
        let edge = find_deferrable_edge(&cycle_strs, &edge_dicts);
        if let Some(edge_obj) = edge.as_object() {
            let from = edge_obj
                .get("from")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let local_field = edge_obj
                .get("localField")
                .and_then(|v| v.as_str())
                .unwrap_or("");

            if let Some(scope_fk) = scope_field_by_model.get(from) {
                if let Some(db_table) = table_map.get(from) {
                    let col_map = column_maps.get(from).cloned().unwrap_or_default();
                    let db_fk_col = col_map
                        .get(local_field)
                        .cloned()
                        .unwrap_or_else(|| local_field.to_string());
                    let db_scope_col = col_map
                        .get(scope_fk)
                        .cloned()
                        .unwrap_or_else(|| scope_fk.clone());
                    let sql = format!(
                        "UPDATE {} SET {} = NULL WHERE {} = {}",
                        dialect.quote_id(db_table),
                        dialect.quote_id(&db_fk_col),
                        dialect.quote_id(&db_scope_col),
                        dialect.param(1)
                    );
                    executor
                        .query(&sql, Some(&[Value::String(scope_value.to_string())]))
                        .await?;
                }
            }
        }
    }

    // Partition sorted nodes: those that depend on cycle nodes must be deleted
    // BEFORE cycles, those that cycle nodes depend on must be deleted AFTER.
    let cycle_node_set: HashSet<&str> = cycles.iter().flatten().map(|s| s.as_str()).collect();

    if !cycle_node_set.is_empty() {
        // Build dependency map: node → set of nodes it depends on
        let mut depends_on: HashMap<&str, HashSet<&str>> = HashMap::new();
        for edge in &schema.edges {
            if edge.from_model != edge.to_model {
                depends_on.entry(edge.from_model.as_str()).or_default().insert(edge.to_model.as_str());
            }
        }

        // Mark nodes that transitively depend on cycle nodes
        let mut depends_on_cycle: HashSet<&str> = HashSet::new();
        for node in &sorted_models {
            if let Some(deps) = depends_on.get(node.as_str()) {
                for dep in deps {
                    if cycle_node_set.contains(dep) || depends_on_cycle.contains(dep) {
                        depends_on_cycle.insert(node.as_str());
                        break;
                    }
                }
            }
        }

        let cycle_dependents: Vec<&str> = sorted_models.iter()
            .filter(|n| depends_on_cycle.contains(n.as_str()))
            .map(|s| s.as_str()).collect();
        let cycle_deps: Vec<&str> = sorted_models.iter()
            .filter(|n| !depends_on_cycle.contains(n.as_str()))
            .map(|s| s.as_str()).collect();

        for model in cycle_dependents.iter().rev() {
            if scope_root_model.as_deref() == Some(*model) {
                continue;
            }
            delete_model(executor, dialect, table_map, column_maps, model,
                scope_value, &scope_field_by_model, refs, schema).await?;
        }

        for cycle in &cycles {
            for model in cycle {
                delete_model(executor, dialect, table_map, column_maps, model,
                    scope_value, &scope_field_by_model, refs, schema).await?;
            }
        }

        for model in cycle_deps.iter().rev() {
            if scope_root_model.as_deref() == Some(*model) {
                continue;
            }
            delete_model(executor, dialect, table_map, column_maps, model,
                scope_value, &scope_field_by_model, refs, schema).await?;
        }
    } else {
        for model in sorted_models.iter().rev() {
            if scope_root_model.as_deref() == Some(model.as_str()) {
                continue;
            }
            delete_model(executor, dialect, table_map, column_maps, model,
                scope_value, &scope_field_by_model, refs, schema).await?;
        }
    }

    // Delete scope root last
    if let Some(ref root) = scope_root_model {
        if let Some(db_table) = table_map.get(root) {
            let col_map = column_maps.get(root).cloned().unwrap_or_default();
            // Bug 4: use dynamic PK field from schema (composite PK: prefer "id")
            let root_model_info = schema.models.iter().find(|m| m.name == *root);
            let root_id_fields: Vec<&crate::types::FieldInfo> = root_model_info
                .map(|mi| mi.fields.iter().filter(|f| f.is_id).collect())
                .unwrap_or_default();
            let root_pk_field_name = root_id_fields.iter()
                .find(|f| f.name.eq_ignore_ascii_case("id"))
                .or(root_id_fields.first())
                .map(|f| f.name.as_str())
                .unwrap_or("id");
            let id_col = col_map
                .get(root_pk_field_name)
                .cloned()
                .unwrap_or_else(|| root_pk_field_name.to_string());
            let sql = format!(
                "DELETE FROM {} WHERE {} = {}",
                dialect.quote_id(db_table),
                dialect.quote_id(&id_col),
                dialect.param(1)
            );
            executor
                .query(&sql, Some(&[Value::String(scope_value.to_string())]))
                .await?;
        }
    }

    Ok(())
}

#[allow(clippy::too_many_arguments)]
async fn delete_model(
    executor: &dyn SqlExecutor,
    dialect: &dyn Dialect,
    table_map: &HashMap<String, String>,
    column_maps: &HashMap<String, HashMap<String, String>>,
    model: &str,
    scope_value: &str,
    scope_field_by_model: &HashMap<String, String>,
    refs: Option<&Value>,
    schema: &SchemaInfo,
) -> Result<(), String> {
    let db_table = match table_map.get(model) {
        Some(t) => t,
        None => return Ok(()),
    };
    let col_map = column_maps.get(model).cloned().unwrap_or_default();

    // Bug 4: Find actual PK field name from schema
    // When multiple is_id fields exist (composite PK), prefer the one named "id"
    let model_info = schema.models.iter().find(|m| m.name == model);
    let id_fields: Vec<&crate::types::FieldInfo> = model_info
        .map(|mi| mi.fields.iter().filter(|f| f.is_id).collect())
        .unwrap_or_default();
    let pk_field_name = id_fields.iter()
        .find(|f| f.name.eq_ignore_ascii_case("id"))
        .or(id_fields.first())
        .map(|f| f.name.as_str())
        .unwrap_or("id");

    if let Some(scope_fk) = scope_field_by_model.get(model) {
        let db_col = col_map
            .get(scope_fk)
            .cloned()
            .unwrap_or_else(|| scope_fk.clone());
        let sql = format!(
            "DELETE FROM {} WHERE {} = {}",
            dialect.quote_id(db_table),
            dialect.quote_id(&db_col),
            dialect.param(1)
        );
        executor
            .query(&sql, Some(&[Value::String(scope_value.to_string())]))
            .await?;
    } else if let Some(ref_records) = refs.and_then(|r| r.get(model)).and_then(|v| v.as_array()) {
        // Bug 3: Accept any non-null value for IDs (not just strings)
        let ids: Vec<Value> = ref_records
            .iter()
            .filter_map(|r| r.get(pk_field_name))
            .filter(|v| !v.is_null())
            .cloned()
            .collect();
        if !ids.is_empty() {
            let id_col = col_map
                .get(pk_field_name)
                .cloned()
                .unwrap_or_else(|| pk_field_name.to_string());
            let placeholders: Vec<String> =
                (0..ids.len()).map(|i| dialect.param(i + 1)).collect();
            let sql = format!(
                "DELETE FROM {} WHERE {} IN ({})",
                dialect.quote_id(db_table),
                dialect.quote_id(&id_col),
                placeholders.join(", ")
            );
            executor.query(&sql, Some(&ids)).await?;
        }
    }

    Ok(())
}

File: sdks/php/src/Teardown.php
<?php

namespace Autonoma\Sdk;

use Autonoma\Sdk\Dialect\DialectInterface;
use Autonoma\Sdk\Types\SQLExecutorInterface;
use Autonoma\Sdk\Types\SchemaInfo;

class Teardown
{
    /**
     * Compute the teardown order for models (reverse topological order).
     * Returns the order, scope root model, cycles, and scope field map.
     *
     * @return array{order: string[], scopeRootModel: string|null, cycles: array[], scopeFieldByModel: array<string, string>}
     */
    public static function computeTeardownOrder(SchemaInfo $schema): array
    {
        // Convert edges to dict format for graph module
        $edgeDicts = [];
        foreach ($schema->edges as $e) {
            $edgeDicts[] = [
                'from' => $e->fromModel,
                'to' => $e->toModel,
                'localField' => $e->localField,
                'foreignField' => $e->foreignField,
                'nullable' => $e->nullable,
            ];
        }

        // Find scope root model
        $scopeRootModel = null;
        foreach ($schema->edges as $edge) {
            if (strtolower($edge->localField) === strtolower($schema->scopeField) &&
                $edge->toModel !== $edge->fromModel) {
                $scopeRootModel = $edge->toModel;
                break;
            }
        }

        // Build map: model -> FK field pointing to scope root
        $scopeFieldByModel = [];
        if ($scopeRootModel !== null) {
            foreach ($schema->edges as $edge) {
                if ($edge->toModel === $scopeRootModel && $edge->fromModel !== $scopeRootModel) {
                    $scopeFieldByModel[$edge->fromModel] = $edge->localField;
                }
            }
        }

        $modelNames = array_map(fn($m) => $m->name, $schema->models);
        $result = Graph::topoSort($modelNames, $edgeDicts);
        $sortedModels = $result['sorted'];
        $cycles = $result['cycles'];

        // Build condensation graph
        $components = [];
        $nodeToComp = [];

        foreach ($cycles as $cycle) {
            $idx = count($components);
            $components[] = $cycle;
            foreach ($cycle as $node) {
                $nodeToComp[$node] = $idx;
            }
        }
        foreach ($sortedModels as $node) {
            $nodeToComp[$node] = count($components);
            $components[] = [$node];
        }

        // Build condensation DAG edges
        $condAdj = [];
        $condInDeg = [];
        for ($i = 0; $i < count($components); $i++) {
            $condAdj[$i] = [];
            $condInDeg[$i] = 0;
        }
        foreach ($schema->edges as $edge) {
            if ($edge->fromModel === $edge->toModel) continue;
            $fc = $nodeToComp[$edge->fromModel] ?? null;
            $tc = $nodeToComp[$edge->toModel] ?? null;
            if ($fc !== null && $tc !== null && $fc !== $tc && !isset($condAdj[$tc][$fc])) {
                $condAdj[$tc][$fc] = true;
                $condInDeg[$fc]++;
            }
        }

        // Kahn's algorithm on the condensation DAG
        $condQueue = [];
        foreach ($condInDeg as $idx => $deg) {
            if ($deg === 0) $condQueue[] = $idx;
        }
        sort($condQueue);
        $condOrder = [];
        while (!empty($condQueue)) {
            sort($condQueue);
            $idx = array_shift($condQueue);
            $condOrder[] = $idx;
            foreach (array_keys($condAdj[$idx]) as $neighbor) {
                $condInDeg[$neighbor]--;
                if ($condInDeg[$neighbor] === 0) {
                    $condQueue[] = $neighbor;
                }
            }
        }

        // Flatten in reverse condensation order, excluding scope root
        $order = [];
        foreach (array_reverse($condOrder) as $compIdx) {
            foreach ($components[$compIdx] as $model) {
                if ($model !== $scopeRootModel) {
                    $order[] = $model;
                }
            }
        }

        return [
            'order' => $order,
            'scopeRootModel' => $scopeRootModel,
            'cycles' => $cycles,
            'scopeFieldByModel' => $scopeFieldByModel,
        ];
    }

    /**
     * Delete all data scoped to scopeValue in reverse topological order.
     *
     * @param string[] $skipModels Models to skip (handled by factory teardown)
     */
    public static function teardown(
        SQLExecutorInterface $executor,
        DialectInterface $dialect,
        array $tableMap,
        array $columnMaps,
        SchemaInfo $schema,
        string $scopeValue,
        ?array $refs = null,
        array $skipModels = [],
    ): void {
        $teardownInfo = self::computeTeardownOrder($schema);
        $order = $teardownInfo['order'];
        $scopeRootModel = $teardownInfo['scopeRootModel'];
        $cycles = $teardownInfo['cycles'];
        $scopeFieldByModel = $teardownInfo['scopeFieldByModel'];

        // Convert edges to dict format for findDeferrableEdge
        $edgeDicts = [];
        foreach ($schema->edges as $e) {
            $edgeDicts[] = [
                'from' => $e->fromModel,
                'to' => $e->toModel,
                'localField' => $e->localField,
                'foreignField' => $e->foreignField,
                'nullable' => $e->nullable,
            ];
        }

        $executor->transaction(function (SQLExecutorInterface $tx) use (
            $dialect, $tableMap, $columnMaps, $scopeRootModel,
            $scopeFieldByModel, $order, $cycles, $edgeDicts,
            $scopeValue, $refs, $schema, $skipModels,
        ) {
            // Break cycles by nullifying deferrable FKs
            foreach ($cycles as $cycle) {
                $edge = Graph::findDeferrableEdge($cycle, $edgeDicts);
                if ($edge !== null) {
                    $scopeFk = $scopeFieldByModel[$edge['from']] ?? null;
                    if ($scopeFk !== null) {
                        $dbTable = $tableMap[$edge['from']] ?? null;
                        $colMap = $columnMaps[$edge['from']] ?? [];
                        if ($dbTable !== null) {
                            $dbFkCol = $colMap[$edge['localField']] ?? $edge['localField'];
                            $dbScopeCol = $colMap[$scopeFk] ?? $scopeFk;
                            $tx->query(
                                sprintf(
                                    'UPDATE %s SET %s = NULL WHERE %s = %s',
                                    $dialect->quoteId($dbTable),
                                    $dialect->quoteId($dbFkCol),
                                    $dialect->quoteId($dbScopeCol),
                                    $dialect->param(1),
                                ),
                                [$scopeValue],
                            );
                        }
                    }
                }
            }

            // Delete in reverse condensation order (dependents first), skipping factory-teardown models
            foreach ($order as $model) {
                if (in_array($model, $skipModels, true)) continue;
                self::deleteModel(
                    $tx, $dialect, $tableMap, $columnMaps, $model,
                    $scopeValue, $scopeFieldByModel, $refs, $schema,
                );
            }

            // Delete scope root last (unless skipped by factory teardown)
            if ($scopeRootModel === null || in_array($scopeRootModel, $skipModels, true)) {
                return;
            }

            $dbTable = $tableMap[$scopeRootModel] ?? null;
            $colMap = $columnMaps[$scopeRootModel] ?? [];
            if ($dbTable !== null) {
                // Find PK field name for scope root model
                $rootModelInfo = null;
                foreach ($schema->models as $m) {
                    if ($m->name === $scopeRootModel) {
                        $rootModelInfo = $m;
                        break;
                    }
                }
                // Composite PK: prefer field named "id"
                $rootIdFields = [];
                if ($rootModelInfo !== null) {
                    foreach ($rootModelInfo->fields as $f) {
                        if ($f->isId) {
                            $rootIdFields[] = $f;
                        }
                    }
                }
                $rootPkField = null;
                foreach ($rootIdFields as $f) {
                    if (strtolower($f->name) === 'id') {
                        $rootPkField = $f;
                        break;
                    }
                }
                if ($rootPkField === null) {
                    $rootPkField = $rootIdFields[0] ?? null;
                }
                $rootPkFieldName = $rootPkField !== null ? $rootPkField->name : 'id';
                $idCol = $colMap[$rootPkFieldName] ?? $rootPkFieldName;
                $tx->query(
                    sprintf(
                        'DELETE FROM %s WHERE %s = %s',
                        $dialect->quoteId($dbTable),
                        $dialect->quoteId($idCol),
                        $dialect->param(1),
                    ),
                    [$scopeValue],
                );
            }
        });
    }

    private static function deleteModel(
        SQLExecutorInterface $tx,
        DialectInterface $dialect,
        array $tableMap,
        array $columnMaps,
        string $model,
        string $scopeValue,
        array $scopeFieldByModel,
        ?array $refs,
        SchemaInfo $schema,
    ): void {
        $dbTable = $tableMap[$model] ?? null;
        if ($dbTable === null) return;
        $colMap = $columnMaps[$model] ?? [];

        // Find actual PK field name from schema
        $modelInfo = null;
        foreach ($schema->models as $m) {
            if ($m->name === $model) {
                $modelInfo = $m;
                break;
            }
        }
        // When multiple isId fields exist (composite PK), prefer the one named "id"
        $idFields = [];
        if ($modelInfo !== null) {
            foreach ($modelInfo->fields as $f) {
                if ($f->isId) {
                    $idFields[] = $f;
                }
            }
        }
        $pkField = null;
        foreach ($idFields as $f) {
            if (strtolower($f->name) === 'id') {
                $pkField = $f;
                break;
            }
        }
        $pkFieldName = $pkField !== null ? $pkField->name : ($idFields[0]->name ?? 'id');

        $scopeFk = $scopeFieldByModel[$model] ?? null;
        if ($scopeFk !== null) {
            $dbCol = $colMap[$scopeFk] ?? $scopeFk;
            $tx->query(
                sprintf(
                    'DELETE FROM %s WHERE %s = %s',
                    $dialect->quoteId($dbTable),
                    $dialect->quoteId($dbCol),
                    $dialect->param(1),
                ),
                [$scopeValue],
            );
        } elseif ($refs !== null && isset($refs[$model])) {
            $ids = [];
            foreach ($refs[$model] as $r) {
                $id = $r[$pkFieldName] ?? null;
                if ($id !== null) {
                    $ids[] = $id;
                }
            }
            if (!empty($ids)) {
                $idCol = $colMap[$pkFieldName] ?? $pkFieldName;
                $placeholders = implode(', ', array_map(fn($i) => $dialect->param($i + 1), range(0, count($ids) - 1)));
                $tx->query(
                    sprintf(
                        'DELETE FROM %s WHERE %s IN (%s)',
                        $dialect->quoteId($dbTable),
                        $dialect->quoteId($idCol),
                        $placeholders,
                    ),
                    $ids,
                );
            }
        }
    }
}
